File: data/skills/user_d6825a13__gcw-webclone/files/scripts/url_safety.py
"""Shared URL boundary checks for GCW Python tools."""

from __future__ import annotations

import re
from urllib.parse import parse_qsl, urlparse


SENSITIVE_QUERY_KEY = re.compile(
    r"(?:^|[-_])(?:key|api[-_]?key|access[-_]?token|token|auth|authorization|signature|sig|secret|password|passwd|credential|code|session|jwt)$",
    re.I,
)


def validate_public_url(value: str, label: str) -> None:
    parsed = urlparse(value)
    if parsed.scheme not in {"http", "https"} or not parsed.netloc or not parsed.hostname:
        raise ValueError(f"{label} must be an absolute http(s) URL")
    if parsed.username or parsed.password:
        raise ValueError(f"{label} must not contain embedded credentials")
    if any(SENSITIVE_QUERY_KEY.search(key) for key, _ in parse_qsl(parsed.query, keep_blank_values=True)):
        raise ValueError(f"{label} must not contain credential-like query parameters")


def validate_relative_route(value: str, label: str) -> None:
    parsed = urlparse(value)
    if parsed.scheme or parsed.netloc:
        raise ValueError(f"{label} must stay on the configured origin")
    if any(SENSITIVE_QUERY_KEY.search(key) for key, _ in parse_qsl(parsed.query, keep_blank_values=True)):
        raise ValueError(f"{label} must not contain credential-like query parameters")

File: data/skills/user_d6825a13__gcw-webclone/files/scripts/url_safety.mjs
const SENSITIVE_QUERY_KEYS = /(?:^|[-_])(?:key|api[-_]?key|access[-_]?token|token|auth|authorization|signature|sig|secret|password|passwd|credential|code|session|jwt)$/i;

export function hasSensitiveQuery(url) {
  return [...url.searchParams.keys()].some((key) => SENSITIVE_QUERY_KEYS.test(key));
}

export function parsePublicHttpUrl(value, label = "URL", { allowSensitiveQuery = false } = {}) {
  let url;
  try {
    url = new URL(value);
  } catch {
    throw new Error(`${label} must be an absolute http(s) URL`);
  }
  if (!["http:", "https:"].includes(url.protocol) || !url.hostname) {
    throw new Error(`${label} must be an absolute http(s) URL`);
  }
  if (url.username || url.password) {
    throw new Error(`${label} must not contain embedded credentials`);
  }
  if (!allowSensitiveQuery && hasSensitiveQuery(url)) {
    throw new Error(`${label} must not contain credential-like query parameters`);
  }
  return url;
}

export function sanitizeUrl(value) {
  try {
    const url = new URL(value);
    url.username = url.username ? "REDACTED" : "";
    url.password = url.password ? "REDACTED" : "";
    for (const key of [...url.searchParams.keys()]) {
      if (SENSITIVE_QUERY_KEYS.test(key)) url.searchParams.set(key, "REDACTED");
    }
    url.hash = "";
    return url.href;
  } catch {
    return value;
  }
}

export function sanitizeText(value) {
  return String(value).replace(/https?:\/\/[^\s"'<>]+/gi, (match) => {
    const trailing = match.match(/[),.;:]+$/)?.[0] || "";
    const url = trailing ? match.slice(0, -trailing.length) : match;
    return `${sanitizeUrl(url)}${trailing}`;
  });
}

export function joinSameOrigin(baseValue, route, label = "route") {
  const base = parsePublicHttpUrl(baseValue, "base URL");
  if (typeof route !== "string" || !route.trim()) throw new Error(`${label} must be a non-empty string`);
  const target = new URL(route, base.href.endsWith("/") ? base.href : `${base.href}/`);
  if (target.origin !== base.origin) throw new Error(`${label} must stay on ${base.origin}`);
  return target.href;
}

export async function assertSameOriginRedirects(value, allowedOrigin, label = "URL", timeoutMs = 30000) {
  let current = parsePublicHttpUrl(value, label);
  for (let hop = 0; hop < 10; hop += 1) {
    const response = await fetch(current, { method: "GET", redirect: "manual", signal: AbortSignal.timeout(timeoutMs) });
    await response.body?.cancel();
    if (response.status < 300 || response.status >= 400) return current.href;
    const location = response.headers.get("location");
    if (!location) return current.href;
    const target = new URL(location, current);
    if (target.origin !== allowedOrigin) throw new Error(`${label} redirect left the allowed origin`);
    current = parsePublicHttpUrl(target.href, `${label} redirect`);
  }
  throw new Error(`${label} exceeded 10 redirects`);
}
